File: src/main/org/nlogo/agent/RootsTable.java
package org.nlogo.agent;

// table lookup of square roots

strictfp class RootsTable {

  private final double[] rootsTable;

  RootsTable(int worldWidth, int worldHeight) {
    rootsTable = new double[worldWidth * worldWidth + worldHeight * worldHeight];
    for (int i = 0; i < rootsTable.length; i++) {
      rootsTable[i] = StrictMath.sqrt(i);
    }
  }

  double gridRoot(double val) {
    int intVal = (int) val;
    if (val == intVal) {
      return gridRoot(intVal);
    } else {
      return StrictMath.sqrt(val);
    }
  }

  double gridRoot(int val) {
    return val < rootsTable.length
        ? rootsTable[val]
        : StrictMath.sqrt(val);
  }

}

File: src/main/org/nlogo/agent/RootsTable.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.agent

// table lookup of square roots

@annotation.strictfp
class RootsTable(worldWidth: Int, worldHeight: Int) {
  private val rootsTable =
    Array.tabulate(worldWidth * worldWidth + worldHeight * worldHeight)(
      StrictMath.sqrt(_))
  def gridRoot(x: Double): Double = {
    val i = x.toInt
    if(x == i)
      gridRoot(i)
    else
      StrictMath.sqrt(x)
  }
  def gridRoot(n: Int): Double =
    if(n < rootsTable.length)
      rootsTable(n)
    else
      StrictMath.sqrt(n)
}
